### ISSUE TITLE
Application Library Final Touches

### ISSUE THREAD

**Issue by WatchMkr** [pop-os/gnome-shell-extension-pop-cosmic#228]
- [x] Allow arrows navigation to scroll in the icon app grid
- [x] PgUp/Dn for moving between folders
- [ ] Dropping an app icon on Create Folder opens the create folder dialog (DEFERRED)
- [x] Center the folders at the bottom of the window
- [x] Open Application Library > Click Icon in the Dock > Library should close
- [x] Open Application Library > Use Super D to open Workspaces > Library should close
- [x] Open Applications Library > Open Launcher > Application Library should close
- [x] Applications button in the Top Bar should stay active when the Application Library is open
- [x] Remove App icon open indicator dot
- [x] Light mode styling (like Launcher)
- [x] Try to improve search result speed to match the old app wall (adjust styling to match mockups if we can improve speed)

**Comment by leviport** [comment]
- [x] Dialogs are fairly narrow, but grow if I type a long folder name:
![Peek 2021-11-10 14-10](https://user-images.githubusercontent.com/13512097/141196243-501fdf56-91a5-44f8-9d2e-d168fca83e79.gif)

- [x] Element focus feels a bit strange in dialogs. Create and Edit dialogs start with text box focused, and Enter does not confirm them. Using Tab or arrow keys gets me to the buttons below where I can use Enter, but then I can't Tab or arrow back up to the text box afterwards (have to click with mouse if I want to edit the text again before choosing)

**Comment by leviport** [comment]
- [ ] Opening App library, then changing workspaces does not close App library (should it?)

**Comment by bflanagin** [comment]
- [x] Recently installed applications initially appear in library, but disappear from the list once the application library is toggled. Happens with both package types.

**Comment by leviport** [comment]
- [ ] Opening App library, then opening Launcher with Super key does not close App library (probably should, right?). 
    - [ ]  If it shouldn't close the app library, then launching an app from Launcher afterwards probably should. Otherwise the App library is sitting in front of the app I just launched, which feels strange.

**Comment by maria-komarova** [comment]
To add to the above, 

- [x] The styling of the search bar should match the styling of the search bar inside the Launcher for consistency - corner radius, background, highlight color, dropshadows (or their absence). 
- [x] The layout for the search results could follow this mockup to keep things a bit more consistent: 
![app-grid-search-results](https://user-images.githubusercontent.com/57502897/141197103-bb404aa8-c1dd-4d9b-8317-6e6046512cc4.png)

- [x] Styling for the edit and delete icon buttons inside the folders: match corner radius to icon buttons in the Settings (like back button), can start with 4-6; background on the mockups is #353535, border same color as the border around the inactive search bar (mockups have it at #171717); the button is 42x42px, icon size 16px, same background color on hover as for the folders. The spacing between edit and delete icons is 12px. 

- [ ] Not sure if Selected state for the folder is currently matching some other styling we have but it could be darker for the sake of contrast.  (DEFERRED)

- [x] The copy and layout for the confirmation dialog (after one presses Delete icon button) should be the following:
![confirmation-dialog](https://user-images.githubusercontent.com/57502897/141199531-5377b603-0409-40f3-8cf7-c786b643299d.png)

- [x] The layout for Creating new folder dialog is below: padding above the header "New Folder" is 18px, 18px between the input box and the buttons at the bottom, 24px padding on left and right, input box on the mockup is 330px wide. 
![new-folder-dialog](https://user-images.githubusercontent.com/57502897/141203136-fd654dab-f5d9-4732-afbf-4cfc11233b8d.png)

- [x] The behavior when one is typing something long can probably match the New Folder input box in Files, where the typed text disappears to the left but the input box stays constrained. The difference should be that it should have Untitled placeholder text and if someone presses Create without typing the name, then the folder should get "Untitled" as the name. 

- [ ] Could we make the transparency gradient at the bottom of the app icons towards the divider like on the mockups to indicate scrolling instead of the scrollbar? If not, we should try adjusting the styling of the to be closer to the background color. (DEFERRED)

**Comment by maria-komarova** [comment]
> I can create multiple folders with the same name. I'm guessing we'd rather prevent this even though it's not a serious bug:

@leviport I actually don't think it is really an issue, having multiple folders of the same name should probably be possible since it doesn't actually lead to any errors or mistakes. So it might be just someone's choice?

**Comment by leviport** [comment]
That's a good point, even though it feels strange lol. I'll edit out that issue. But I also found another one: 

- [x] Folders can have no title. They should probably default to "Untitled" or something like that. Also, would we want to prevent folder names that are just whitespace?
![2021-11-10_15-13](https://user-images.githubusercontent.com/13512097/141202806-3a9246bd-5674-431b-a2f3-c8be91be25a5.png)

**Comment by maria-komarova** [comment]
Just updated my comment above. If there is no name typed, then the Folder should get "Untitled" as the name.

**Comment by 13r0ck** [comment]
# Features
Can `PgUp/Dn for moving between folders` be done via a keyboard shortcut? I would like the option to change it to Ctrl+Tab

# Bugs
1. Using the gnome typing booster `ibus-typing-booster` and selecting an option causes the library to freeze (highlight and clicking do nothing).

https://user-images.githubusercontent.com/58987761/141213897-f104bf54-5fe6-45ee-b6a1-406f6e41164f.mp4

2. After renaming a folder, the name in the folder changes, but not the title (until folder is re-opened)

https://user-images.githubusercontent.com/58987761/141213895-e5243004-ca0c-4451-bb45-4a66fea045c4.mp4

3. If the calendar view is opened, opening application library and/or launcher/workspaces doesn't close it

https://user-images.githubusercontent.com/58987761/141213891-5ac208f4-ccdd-4741-80a3-a02369ccea3e.mp4

> Note: Scaling is awkward on small screens. QA already had a conversation with Maria

**Comment by ids1024** [comment]
> Can PgUp/Dn for moving between folders be done via a keyboard shortcut? I would like the option to change it to Ctrl+Tab

Possible, though shortcuts there are normally universal ones that can be used everywhere.

It also probably wouldn't make sense without a fix for https://github.com/pop-os/gnome-control-center/issues/142, since it wouldn't be possible to manually set a binding like `PageUp` without a modifier.

**Comment by 13r0ck** [comment]
maybe in a config file? :man_shrugging:

**Comment by 13r0ck** [comment]
When using two monitors it it possible to make the folders overlap into the next screen (This is visible on the actual display, not just the screenshot.)
![Screenshot from 2021-11-11 10-56-32](https://user-images.githubusercontent.com/58987761/141346985-c0b7a568-e9ab-414e-b25a-9b06e7bd1756.png)

Creating a bunch of folders on one screen will then get cut off on another (This fits into the scaling issue that Maria and QA talked about yesterday.)
![Screenshot from 2021-11-11 10-54-52](https://user-images.githubusercontent.com/58987761/141347281-fbb1bf51-825f-450c-98d3-a38eba2cb3d5.png)
![Screenshot from 2021-11-11 10-54-59](https://user-images.githubusercontent.com/58987761/141347397-697cb94e-b79b-448f-99e1-28dda55ba94f.png)

**Comment by ids1024** [comment]
I'm not sure what we want to do about folders overflowing the screen. Earlier we talked about limiting the number of folders, which shouldn't be too hard, but I don''t know how many folders someone is likely to use.

I'm also not sure how much would fit on screen at 720p (I assume that's roughly the lowest we'd need to support?).

**Comment by senntore** [comment]
I'd like to add few things:

- First app in search results should be selected by default, so we can open them with return key right after we enter the app name.
Right now we have to navigate using arrows after the search, before we can open the top result.
![Screenshot-2021-11-12](https://user-images.githubusercontent.com/58325574/141426709-70054827-e931-4518-8503-f891112f66bc.png)
- The Search toggle for Pop Shop's suggestions in Settings>Applications>Pop!_Shop should integrate with Application library unless they are planned to be added in Settings>Desktop>General, though this could be a separate issue for [pop-os/desktop-widget](https://github.com/pop-os/desktop-widget)

Another small issue, are notifications & tooltips(when hovering over an app's icon in dock) supposed to be shown below Application Library's window?
Screenshot:
![Screenshot-2021-11-1](https://user-images.githubusercontent.com/58325574/141445441-9b629740-dda6-48dd-8e69-d2aafcd3a743.png)
![Screenshot-2021-11-12,1](https://user-images.githubusercontent.com/58325574/141445445-111d4ad8-d669-4f29-9d68-3cbbd78440de.png)

**Comment by maria-komarova** [comment]
Good points. Yes, makes sense to have the first search result highlighted so that one can open it right away with the keyboard. 
It might also be better if the tooltip goes above the Application Library window.

**Comment by leviport** [comment]
- [x]  I'm also seeing https://github.com/pop-os/beta/issues/319 which should probably be addressed eventually. I'd expect the app library to close when one of the right click items is clicked.

**Comment by WatchMkr** [comment]
Release blocking list

- [x] Recently installed applications initially appear in library, but disappear from the list once the application library is toggled. Happens with both package types. 
- [x] Allow arrows navigation to scroll in the icon app grid
- [x] Open Application Library > Click Icon in the Dock > Library should close
- [x] Open Application Library > Use Super D to open Workspaces > Library should close
- [ ] Opening App library, then changing workspaces does not close App library
- [x] Light mode styling (like Launcher)
- [x] The styling of the search bar should match the styling of the search bar inside the Launcher for consistency - corner radius, background, highlight color, dropshadows (or their absence). 
- [x] Styling for the edit and delete icon buttons inside the folders: match corner radius to icon buttons in the Settings (like back button), can start with 4-6; background on the mockups is #353535, border same color as the border around the inactive search bar (mockups have it at #171717); the button is 42x42px, icon size 16px, same background color on hover as for the folders. The spacing between edit and delete icons is 12px. 
- [x] Left align "Available to Install" app icons (this is because there is a label. Results should be left aligned under a label)
- [x] In search results, remove the open application indicator dot (like the main view)
- [x] Highlight the first search result (so 'enter' will launch the app)

**Comment by ids1024** [comment]
> Open Application Library > Click Icon in the Dock > Library should close

I think https://github.com/pop-os/cosmic/pull/251 should have fixed this one.

**Comment by WatchMkr** [comment]
> Open Application Library > Click Icon in the Dock > Library should close

To reveal this bug I click the Show Applications icon in the dock then click the icon again.

**Comment by ids1024** [comment]
Ah, I see what you mean now.

**Comment by ids1024** [comment]
> Recently installed applications initially appear in library, but disappear from the list once the application library is toggled. Happens with both package types.

I seem to see installed applications appear correctly. Though if I have an app in a folder, uninstall, and reinstall with the Library open on Library Home, the app appears there but then no longer appears in Library Home when the Library is re-opened, since it should instead be in a folder.

I see how to fix that, but I guess that's not what you're describing?

**Comment by senntore** [comment]
Not sure if the last merge caused this but launching suggestions from Pop Shop(not installed apps) shows Workspaces view along with opening it in Pop Shop.

**Comment by bflanagin** [comment]
> > Recently installed applications initially appear in library, but disappear from the list once the application library is toggled. Happens with both package types.
> 
> I seem to see installed applications appear correctly. Though if I have an app in a folder, uninstall, and reinstall with the Library open on Library Home, the app appears there but then no longer appears in Library Home when the Library is re-opened, since it should instead be in a folder.
> 
> I see how to fix that, but I guess that's not what you're describing?

Not exactly. The application icon is present the first time the Application Library appears after software installation, but when you close the application library and then reopen it the newly installed software's icon is gone. Restarting Gnome fixes the issue. 
 
Steps:
1. Install something from Pop!_Shop
2. Open App Library (Icon should be there)
3. Close App Library
4. ReOpen App Library (Icon is gone)

At least that is what it did at the time, this morning no new application icons appear after install until gnome is restarted.

**Comment by ids1024** [comment]
Ah, okay. I can reproduce that as long as the app wasn't installed before (uninstalling/reinstalling without restarting Gnome Shell doesn't reproduce it), and I see what seems to be the cause.

**Comment by senntore** [comment]
Is there going to be a **No Results** or similar screen when the search result is empty?

Also will we able to toggle **available to install** when it'll be out of beta? 
It takes time to load, the icons are sometimes blurry, the result also vary a lot from the same search in pop shop.
It doesn't have to be available in gui(for now), even just a command would do.

**Comment by 13r0ck** [comment]
@senntore 
> Is there going to be a No Results or similar screen when the search result is empty?

We are currently building ISO's for 21.10's official release. That will not make it in time. Feel free to make an issue in this repo about it, so we can discuss and maybe add it to the list for release on a later date.

> Also will we able to toggle available to install when it'll be out of beta?

Again, not in time for 21.10 official release. Make an issue if you would like to start the discussion about maybe building that feature
